mask.json

```diff
@@ -1,14 +1,15 @@
 {
     "id": "mask",
     "summary": "Apply a mask",
-    "description": "Applies a mask to a raster data cube. A mask can either be specified as:\n\n* **Raster data cube** for which parallel pixels among `data` and `mask` are compared and those pixels in `data` are replaced, which are non-zero (for numbers) or `true` (for boolean values) in `mask`.\n* **GeoJSON or vector data cube** containing one or more polygons. All pixels for which the point at the pixel center intersects with the corresponding polygon (as defined in the Simple Features standard by the OGC) are replaced.\n\nThe pixel values are replaced with the value specified for `replacement`, which defaults to `null` (no data). No data values will be left untouched by the masking operation.",
+    "description": "Applies a mask to a raster data cube. A mask can either be specified as:\n\n* **Raster data cube** for which parallel pixels among `data` and `mask` are compared and those pixels in `data` are replaced, which are non-zero (for numbers) or `true` (for boolean values) in `mask`.\n* **GeoJSON or vector data cube** containing one or more polygons. All pixels for which the point at the pixel center intersects with any polygon (as defined in the Simple Features standard by the OGC) are replaced.\n\nThis behavior can be inverted with the parameter `invert`.\n\nThe pixel values are replaced with the value specified for `replacement`, which defaults to `null` (no data). No data values in `data` or `mask` will be left untouched by the masking operation.",
     "categories": [
         "masks"
     ],
     "parameter_order": [
         "data",
         "mask",
-        "replacement"
+        "replacement",
+        "invert"
```

Comment:
Good point indeed, "invert" does not clearly communicate what the inverted and non-inverted behaviour are.
The nice thing about `active_mask_value` is that it even could be extended to integer valued mask (not only binary masks)

On the other hand, the alternative parameter names you provide only work well for raster based masks. For polygon based masks they make less sense, while "invert" is a lot more intuitive.

Not sure what the best approach is here.
We could even think about splitting the process in a raster based process and a polygon based process. It would make the process descriptions and documentation also simpler (no need for "if raster mask then ..., if polygon then ...." repeatedly).

Reply:
I'll try to split the processes, I think. It would really make the process(es) simpler and we can better describe specifics.

Reply:
Thinking about the raster masking, what would be a good default for `active_mask_value`? Or should it be required? Current behavior is "non-zero (for numbers) or `true` (for boolean values)" is getting replaced. That behavior doesn't work with a parameter with a specific value any longer. I'm leaning towards a default value `true` as it's easy to create a boolean mask with apply(data, gt(x, 0)) for example.

Reply:
I think it's not so uncommon to automatically translate bool `true` to "non-zero" in int land. So if you have an int-valued mask and the default (or explicitly set) value of param `active_mask_value` is bool `true`, you mask out the pixels with non-zero mask values.

Another issue when extending to int-masks with an active_mask_value `M`: do you mask out pixels `where mask is M`  or `where is not M`?
Both use cases are useful, I guess. Take for example a coverage class based mask with classes like "vegetated", "water", "urban", "clouded". In some cases you probably want to keep vegetated pixels and replace all the rest, while in other cases you want to replace clouded pixels and keep all the rest. To do this with an int-mask, you probably still need an "invert"-like parameter to toggle between these two behaviors.

Reply:
> I think it's not so uncommon to automatically translate bool true to "non-zero" in int land. 

Yes, although this is some kind of implicit magic and what happens if a user really just want to replace true... 

> Another issue when extending to int-masks with an active_mask_value M: do you mask out pixels where mask is M or where is not M?

@soxofaan Yes, that's exactly what I'm thinking about at the moment and what keeps me from pushing the changes for review. Another option would be to define a expression like callback where you could use things like eq(), neq(), between(), not(between()) etc. Then we are not restricted to a single value and don't need the invert param.

Reply:
@soxofaan @mkadunc Please re-review the recent changes. There's now mask with an expression and mask_polygon.

Reply:
Quick note: from the standpoint of the python client, the approach with an expression feels like overkill because we already have  very easy, pythonic "band math" to build boolean masks  (e.g. `(coverage != 4) & (coverage != 5)` . I'm not sure if it's good for API orthogonality and simplicity.

Maybe, in the `mask` process,  we should just focus on boolean masks and provide basic support for integer masks? More advanced integer mask use cases are still possible anyway by first computing  a boolean mask and  using that in the mask process.

Reply:
@soxofaan Other clients can't do that though. How do you translatethe Python example into a process graph?

I don't understand how your comments https://github.com/Open-EO/openeo-processes/pull/110#discussion_r365274430 and https://github.com/Open-EO/openeo-processes/pull/110#discussion_r365324900 - they seem to contradict each other?

So at the moment I'm either in favor of a boolean only approach and thus people need to prepare a mask with true values (so they need to create the expression in apply anyway, so what's the benefit?) or having the expressions in mask directly. It seems the other options have too many drawbacks.

Reply:
> How do you translate the Python example into a process graph?

see https://gist.github.com/soxofaan/1121ca95b45fc43f697fc82b7e5ed7b7 for example

> your comments seem to contradict each other

just some samples from a thought process :)  . 
I guess I'm converging to the conclusion to keep things simple and just support a boolean mask (the API has all the functionality already to build these, no need to replicate that inside a process parameter) .
Also note that the original request from #103 was about an invert option for polygon based masking, not raster masks.

Reply:
For me it seems that having the expression is simpler for a user though. In the end it's just an implicit apply call that a user would need to specify explicitly for a mask.

Reply:
In R for raster masking, we have two values: what value in the mask we care about, and what value we replace those pixels with in the raster we're masking. So we just have two integer (arrays). The defaults can be 1 and NA respectively, because normally users want to put NAs into areas that they want to mask.
For vector masking, same idea but you can have a parameter that says `replace_inside=TRUE`, that way values inside of the geometry will get replaced, otherwise values outside the geometry will be replaced.

Reply:
Telco: Remove expressions for now.